### ISSUE TITLE
Add support for degradation modes

### ISSUE THREAD

**Issue by ejfdickinson** [FaradayInstitution/BPX#60]
Consider adding support within BPX for standard degradation modes that can be expressed within a physics-based model:

- Loss of (Accessible) Active Material (LAM)
- Loss of (Cyclable) Lithium Inventory (LLI)

**Comment by RobertTrieblAVL** [comment]
I agree that standardization of LAM and LLI makes sense, though we need to agree on the quantity which is represented. This can be charge (As) or relative charge (%)
Regarding LAM (cathode or anode), this can be expressed as 

- Volume fraction (example: LAM cathode is the volume fraction of active cathode material which is inactive) I would prefer this option
- Total in As which is inactive (example: LAM cathode is 150 As).
- For multi-material: should this be specified for each material?

Regarding LLI, this is even trickier, as LLI corresponds to a relative shift between cathode and anode OCP. I see following options:

- Charge lost due to LLI in As
- Charge lost relative to nominal battery capacity (charge between soc=0 and soc=1)
- Charge lost relative to anode charge between theta_min and theta_max
- Charge lost relative to cathode charge between theta_min and theta_max

**Comment by ejfdickinson** [comment]
@RobertTrieblAVL

We would conventionally define LAM and LLI relative to the properties of "active material [content]" and "[cyclable] lithium inventory" for the fresh cell, where the latter is the reference performance parameter set in the BPX for an "un-aged" cell (LAM_m = LLI = 0), typically the cell after formation.

Active material content ratios must equate to active material volume ratios and, due to the spherical particle definition in BPX, also to volumetric surface area ratios.

Cyclable Li inventory is the sum of the mol Li inserted in all active materials (neg and pos) at any given SOC, and must equate for all SOCs according the beginning-of-life parameterisation.

As usual I had some struggles with LaTeX in GitHub Markdown but these are the equations:

1 - LAM(material) = (active material volume fraction, material) / (active material volume fraction, material, BOL)

1 - LLI = (mol Li inserted in negative electrode at EOC + mol Li inserted in positive electrode at EOC) / (mol Li inserted in negative electrode at EOC/BOL + mol Li inserted in positive electrode at EOC/BOL)

BOL = beginning of life ("fresh cell" after formation)
EOC = end of charge

**Comment by ejfdickinson** [comment]
@rtimms @ikorotkin 

I started to develop some proposed specification for degradation modes. Please let me know if you have any comments so far.

- We can decide on whether to segregate degradation variables in their own section (under `Parameterisation`, or `State` with the environment variables) or include them the appropriate parameterisation sections to which they apply.
  - I think it is much clearer to consider the degradation variables as defining an altered 'state' of the parameterised cell as expressed independently in `Parameterisation`. One can always choose to define the parameterisation explicitly and manually for an aged cell, in which case the degradation state would start to become ambiguous, so seeing the degradation state as relative to the parameterisation makes most sense to me.

__Proposal__:

## Degradation state

Degradation state can optionally be specified in the `[State] > [Degradation]` node.

Degradation state comprises:

- thermodynamic degradation state parameters (LAM/LLI)
- degradation factors

When specifying the degradation state, the contents of `[Parameterisation]` are considered to define a _reference state_ of the non-degraded (unaged) cell. Degradation state specification implies preprocessing of the parameterisation to yield a new parameter set for use in a simulation of a degraded (aged) cell.

### Thermodynamic degradation state parameters

The following thermodynamic degradation state parameters are supported:

- Loss of (accessible) active material (LAM)
- Loss of (cyclable) lithium inventory (LLI)

These fields are optional. If absent, the parameterisation is considered equal to the reference state. If present, a full set of LAM (for every electrode/material) and LLI must be specified. Implicitly, the values of LAM/LLI are all 0 for the reference state, but note the **Caution** below.

These can be encoded within `Degradation` as:

`LLI`

and

`LAM: [electrode]`

or 

`LAM: [material]`

where `electrode` and `material` are keys associated with a set of electrode (material) properties.

LLI and LAM modify other parameters according to the following rules, where the subscript $_,\mathrm{ref}$ indicates the value in the reference state:

$$1-\mathrm{LAM} = \frac{b_{k,m}}{b_{k,m,\mathrm{ref}}}$$

$$1-\mathrm{LLI} = \frac{n_\mathrm{Li,tot}}{n_\mathrm{Li,tot,ref}}$$

```math
n_\mathrm{Li,tot} =L_\mathrm{p}\sum_m{\theta^\mathrm{max}_{\mathrm{p},m}c^\mathrm{max}_{\mathrm{p},m}\phi_{\mathrm{p},m}} + L_\mathrm{n}\sum_q{\theta^\mathrm{min}_{\mathrm{n},q}c^\mathrm{max}_{\mathrm{n},q}\phi_{\mathrm{n},q}}
```
$$\phi_{k,m} = \frac{b_{k,m}R_{k,m}}{3}$$

for all materials enumerated $m$ in the positive electrode and $q$ in the negative electrode.

Specifying thermodynamic degradation state parameters causes the stoichiometry limits to be recomputed with the revised material contents, based on the following rules.

**Caution**: The handling of thermodynamic degradation state parameters requires that stoichiometry limits in the BPX file are defined in a thermodynamically self-consistent manner for the reference state. If the stoichiometry limits specified in the reference state are not thermodynamically self-consistent, there will be a discontinuous difference between the parameterisation expressed without a degradation state and the parameterisation expressed when LAM/LLI are present but equal to zero.

TO WRITE

### Degradation factors

Beside thermodynamic degradation state parameters, a degradation factor can be declared for any property (except OCP, stoichiometry):

`Degradation factor: [colon-separated json hierarchy]`

Degradation factors are optional and have an implied value of 1 if not specified. The degradation factor acts as a direct scalar multiplier to the corresponding quantity.
